### ISSUE TITLE
Should constant matchers be value matchers

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#1406]
See [pattern proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md#constant-matcher).

The constant matcher can be used to match specific values. It must be a constant value. This makes sense for when it's used in a `switch` `case` matcher where we currently require expressions to be constant, but not necessarily in a more dynamic check like `e is matcher`.

Example:
```dart
Iterable<List<T>> listsWhichStartWith<T>(T value, Iterable<List<T>> lists) sync* {
  for (var p in lists) if (p is [value, ... var rest]) yield rest;  
}
```
(assuming the `List` rest-matcher TODO)

If we only allow constant values to be used as matchers, we won't be able to do parameterized pattern matching on provided values. I think we should allow that. Not necessarily in `switch`, but in `is` and `as` checks.

**Comment by leafpetersen** [comment]
I'm not sure what the proposal here is.  @munificent seems to be suggesting in discussion in #2484 that this is about allowing general expressions in patterns, but I'm not sure I see the connection, at least directly.   Put another way, what is a "value matcher"?

**Comment by lrhn** [comment]
I'm asking whether the values matched for equality necessarily have to be constant.

Consider 
```dart
bool startsWith<T>(T value, List<T> list) {
  if (list case [value, ...]) {
    return true;
  } else {
    return false;
  }
}
```
(Which I know is so simple it's silly.)

Nothing prevents allowing a non-constant expression, the only operations we do on the value is to compare for equality.

It would mean that we have to specify *when* the expression is evaluated, since it can have side effects.

We could restrict it to side-effect free expressions (something like potentially constant expressions where the variables can be any local variable). Then side effects are not an issue.
That would probably mean that we should introduce "side effect free computation" as a concept in general, which would be complicated.

Also, the above can be written as
```dart
  if (list case [var x, ...] when x == value)
```
so it's not a big loss to not be able to do the non-constant match inline in the pattern.

**Comment by leafpetersen** [comment]
@lrhn I think what's not clear to me is what you mean by "expressions".  The example you use is just a variable (I think), which is fine syntactically - that's what I suggested as well in #2484 .  But if you really mean expressions, how do you want to handle the syntax?  That is, how do I know whether `var x = 1; var y = 2; var l = [1, 2]; switch (l) case [x, y]:..` is asking the question:
  -  "is `l` a List, and if so, are `x == l[0]` and `y = l[1]` both true?", which is what the current semantics specify
    -  The answer to this is "yes" 
  - is `[x, y] == l` true, which is what I think a value matcher syntax with general expressions would give.
     - The answer to this is "false"

For example, one way to do resolve this would be to use `new` as an inclusion:  `new e` is a pattern which matches a scrutinee `s` iff `e` evaluates to a value `v` (possibly with caching) and `v == s`.  This doesn't give the nice variable syntax that you use above though.

> Also, the above can be written as
> 
> ```dart
>   if (list case [var x, ...] when x == value)
> ```

Nit: `value == x`.

**Comment by lrhn** [comment]
Good point. If we allow general expressions as pattern "value matches", then we get ambiguity between collection literal values and collection destructuring patterns. For `const` values, we need to add a `const` in front, we don't get that for values.

On the other hand, allowing some expressions, but not other, is not great for usability. 

We probably should not try to extend to allowing arbitrary expressions as value matchers.

We *can* allow arbitrary expressions after `==`, to be evaluated at the time of the `==` check.
Can even allow it after relational patterns like `< e`, that just prevents static exahustability analysis.

Not sure we need to, so we should definitely not try it until someone actually has a use-case.

**Comment by munificent** [comment]
It sounds like there's two orthogonal questions:

1. For the subset of expressions that are valid to appear in a pattern, do we loosen the restriction on them being constant and allow non-constant expressions? In #2484, @leafpetersen is specifically asking about named variable constant patterns. Right now, the only other expressions that can appear directly in patterns is primitive literals which are always const anyway, so #2484 basically covers everything.

2. Should we expand the set of expressions that are syntactically allowed to be used as a pattern? We could, for example, make `1 + 2` a valid pattern (that would match if the value is 3).

For 1, I think the best answer for the initial release is "no". I think it's probably feasible, but I'd want more time to think through the implications. It's easy to relax this restriction in a later release if it turns out to be important. In the meantime, you can always use guards.

For 2, it gets hard quickly. The pattern grammar is already pretty rich and, by design, it heavily mirrors the expression grammar. You can always access the full expression grammar (though the result must be a valid constant expression) using an `==` pattern, so I don't think it buys much.

I'm going to move this into "patterns-later" but feel free to move it back if you think it's important to consider in the initial release.
